File: Java/NeuralNetwork.java
import java.util.Random;

class NeuralNetwork {
    private final int[] layerSizes;
    private final double learningRate;
    private final double[][][] weights;
    private final double[][] biases;
    private final Random random = new Random(42);

    public NeuralNetwork(int[] layers, double learningRate) {
        this.layerSizes = layers;
        this.learningRate = learningRate;
        this.weights = new double[layers.length - 1][][];
        this.biases = new double[layers.length - 1][];
        initializeWeightsAndBiases();
    }

    private void initializeWeightsAndBiases() {
        for (int i = 0; i < layerSizes.length - 1; i++) {
            weights[i] = new double[layerSizes[i]][layerSizes[i + 1]];
            biases[i] = new double[layerSizes[i + 1]];
            for (int j = 0; j < layerSizes[i]; j++) {
                for (int k = 0; k < layerSizes[i + 1]; k++) {
                    weights[i][j][k] = (random.nextDouble() * 2) - 1;  // Valores aleatórios entre -1 e 1
                }
            }
            for (int j = 0; j < layerSizes[i + 1]; j++) {
                biases[i][j] = (random.nextDouble() * 2) - 1;
            }
        }
    }

    private double sigmoid(double x) {
        return 1 / (1 + Math.exp(-x));
    }

    private double sigmoidDerivative(double x) {
        return x * (1 - x);
    }

    private double[] feedforward(double[] inputs) {
        double[] activations = inputs;
        for (int i = 0; i < weights.length; i++) {
            double[] nextActivations = new double[layerSizes[i + 1]];
            for (int j = 0; j < layerSizes[i + 1]; j++) {
                nextActivations[j] = biases[i][j];
                for (int k = 0; k < layerSizes[i]; k++) {
                    nextActivations[j] += activations[k] * weights[i][k][j];
                }
                nextActivations[j] = sigmoid(nextActivations[j]);
            }
            activations = nextActivations;
        }
        return activations;
    }

    public int predict(double[] inputs) {
        double[] outputs = feedforward(inputs);
        int maxIndex = 0;
        for (int i = 1; i < outputs.length; i++) {
            if (outputs[i] > outputs[maxIndex]) maxIndex = i;
        }
        return maxIndex;
    }

    public void train(double[][] trainImages, int[] trainLabels, int epochs) {
        for (int epoch = 0; epoch < epochs; epoch++) {
            for (int i = 0; i < trainImages.length; i++) {
                double[] inputs = trainImages[i];
                int label = trainLabels[i];

                // Passo de feedforward para obter todas as ativações de cada camada
                double[][] activations = new double[layerSizes.length][];
                activations[0] = inputs;
                for (int layer = 1; layer < layerSizes.length; layer++) {
                    activations[layer] = new double[layerSizes[layer]];
                    for (int j = 0; j < layerSizes[layer]; j++) {
                        double sum = biases[layer - 1][j];
                        for (int k = 0; k < layerSizes[layer - 1]; k++) {
                            sum += activations[layer - 1][k] * weights[layer - 1][k][j];
                        }
                        activations[layer][j] = sigmoid(sum);
                    }
                }

                // Backpropagation: calcula o erro e atualiza pesos e biases
                double[][] errors = new double[layerSizes.length][];
                errors[layerSizes.length - 1] = new double[layerSizes[layerSizes.length - 1]];
                
                // Erro na camada de saída
                for (int j = 0; j < layerSizes[layerSizes.length - 1]; j++) {
                    double target = (j == label) ? 1.0 : 0.0;
                    errors[layerSizes.length - 1][j] = (target - activations[layerSizes.length - 1][j]) 
                                                        * sigmoidDerivative(activations[layerSizes.length - 1][j]);
                }
                
                // Backpropagation nas camadas ocultas
                for (int layer = layerSizes.length - 2; layer > 0; layer--) {
                    errors[layer] = new double[layerSizes[layer]];
                    for (int j = 0; j < layerSizes[layer]; j++) {
                        double error = 0.0;
                        for (int k = 0; k < layerSizes[layer + 1]; k++) {
                            error += errors[layer + 1][k] * weights[layer][j][k];
                        }
                        errors[layer][j] = error * sigmoidDerivative(activations[layer][j]);
                    }
                }

                // Atualização dos pesos e biases
                for (int layer = 0; layer < weights.length; layer++) {
                    for (int j = 0; j < layerSizes[layer + 1]; j++) {
                        for (int k = 0; k < layerSizes[layer]; k++) {
                            weights[layer][k][j] += learningRate * errors[layer + 1][j] * activations[layer][k];
                        }
                        biases[layer][j] += learningRate * errors[layer + 1][j];
                    }
                }
            }
            System.out.printf("Epoch %d concluído.%n", epoch + 1);
        }
    }

    public double evaluate(double[][] testImages, int[] testLabels) {
        int correct = 0;
        for (int i = 0; i < testImages.length; i++) {
            if (predict(testImages[i]) == testLabels[i]) correct++;
        }
        return (double) correct / testImages.length;
    }
}

File: Python/neural_network.py
import numpy as np

class NeuralNetwork:
    def __init__(self, layers, learning_rate):
        self.layer_sizes = layers          # Tamanho das camadas
        self.learning_rate = learning_rate # Taxa de aprendizado
        self.weights = []                  # Lista de pesos
        self.biases = []                   # Lista de biases
        self.initialize_weights_and_biases()

    # Inicialização de pesos e biases
    def initialize_weights_and_biases(self):
        np.random.seed(42)
        for i in range(len(self.layer_sizes) - 1):
            layer_weights = np.random.uniform(-1, 1, (self.layer_sizes[i], self.layer_sizes[i + 1]))
            layer_biases = np.random.uniform(-1, 1, self.layer_sizes[i + 1])
            self.weights.append(layer_weights)
            self.biases.append(layer_biases)

    # Função de ativação sigmoide e sua derivada
    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))

    def sigmoid_derivative(self, x):
        return x * (1 - x)

    # Feedforward: calcula a saída da rede e armazena ativações de cada camada
    def feedforward(self, inputs):
        activations = [inputs]  # Armazena ativações de cada camada
        for i in range(len(self.weights)):
            inputs = self.sigmoid(np.dot(inputs, self.weights[i]) + self.biases[i])
            activations.append(inputs)
        return activations

    # Previsão da classe para uma entrada
    def predict(self, inputs):
        outputs = self.feedforward(inputs)[-1]  # Pega a última ativação (saída final)
        return np.argmax(outputs)

    # Treinamento com backpropagation
    def train(self, train_images, train_labels, epochs):
        for epoch in range(epochs):
            for inputs, label in zip(train_images, train_labels):
                # Passo de feedforward para obter todas as ativações
                activations = self.feedforward(inputs)
                outputs = activations[-1]  # Saída final da rede

                # Calcula o erro da camada de saída
                output_errors = np.array([(1 if j == label else 0) - outputs[j] for j in range(self.layer_sizes[-1])])

                # Backpropagation para ajustar pesos e biases
                errors = [output_errors]
                for j in range(len(self.weights) - 1, 0, -1):
                    layer_errors = np.dot(errors[0], self.weights[j].T) * self.sigmoid_derivative(activations[j])
                    errors.insert(0, layer_errors)

                # Atualização dos pesos e biases com gradientes
                for j in range(len(self.weights)):
                    self.weights[j] += self.learning_rate * np.outer(activations[j], errors[j])
                    self.biases[j] += self.learning_rate * errors[j]

            print(f"Epoch {epoch + 1} concluído.")

    # Avaliação da rede
    def evaluate(self, test_images, test_labels):
        correct_predictions = sum(1 for inputs, label in zip(test_images, test_labels) if self.predict(inputs) == label)
        return correct_predictions / len(test_images)
